File: pglite/src/pglite-custom/release/runtime/lifecycle.ts
import type { RuntimeModule } from "../core/runtimeTypes.js";

function asArray(
	hooks: RuntimeModule["preRun"] | RuntimeModule["postRun"] | undefined,
): Array<() => void> {
	if (!hooks) {
		return [];
	}
	if (Array.isArray(hooks)) {
		return hooks;
	}
	return [hooks as unknown as () => void];
}

export function mergePreRunHooks(
	moduleArg: Partial<RuntimeModule>,
	hook: () => void,
): void {
	const hooks = asArray(moduleArg.preRun);
	hooks.push(hook);
	moduleArg.preRun = hooks;
}

export function mergePostRunHooks(
	moduleArg: Partial<RuntimeModule>,
	hook: () => void,
): void {
	const hooks = asArray(moduleArg.postRun);
	hooks.push(hook);
	moduleArg.postRun = hooks;
}

File: src/main/java/io/github/hidekatsu_izuno/pglite_jdbc/pglite/release/runtime/lifecycle.java
package io.github.hidekatsu_izuno.pglite_jdbc.pglite.release.runtime;

import io.github.hidekatsu_izuno.pglite_jdbc.pglite.release.core.runtimeTypes;
import java.util.ArrayList;

public class lifecycle {
    private lifecycle() {}

    public static void mergePreRunHooks(runtimeTypes.RuntimeModule moduleArg, Runnable hook) {
        var hooks = new ArrayList<>(moduleArg.preRun());
        hooks.add(hook);
        moduleArg.setPreRun(hooks);
    }

    public static void mergePostRunHooks(runtimeTypes.RuntimeModule moduleArg, Runnable hook) {
        var hooks = new ArrayList<>(moduleArg.postRun());
        hooks.add(hook);
        moduleArg.setPostRun(hooks);
    }
}
